crates/bevy_ecs/src/schedule_v3/executor/mod.rs

```diff
@@ -0,0 +1,87 @@
+mod multi_threaded;
+mod simple;
+mod single_threaded;
+
+pub use self::multi_threaded::MultiThreadedExecutor;
+pub use self::simple::SimpleExecutor;
+pub use self::single_threaded::SingleThreadedExecutor;
+
+use std::cell::RefCell;
+
+use fixedbitset::FixedBitSet;
+
+use crate::{
+    schedule_v3::{BoxedCondition, NodeId},
+    system::BoxedSystem,
+    world::World,
+};
+
+/// Types that can run a [`SystemSchedule`] on a [`World`].
+pub(super) trait SystemExecutor: Send + Sync {
+    fn kind(&self) -> ExecutorKind;
+    fn init(&mut self, schedule: &SystemSchedule);
+    fn run(&mut self, schedule: &mut SystemSchedule, world: &mut World);
+}
+
+/// Specifies how a [`Schedule`](super::Schedule) will be run.
+///
+/// [`MultiThreaded`](ExecutorKind::MultiThreaded) is the default.
+#[derive(PartialEq, Eq, Default)]
+pub enum ExecutorKind {
+    /// Runs the schedule using a single thread.
+    SingleThreaded,
+    /// Like [`SingleThreaded`](ExecutorKind::SingleThreaded) but calls [`apply_buffers`](crate::system::System::apply_buffers)
+    /// immediately after running each system.
+    Simple,
+    #[default]
+    /// Runs the schedule using a thread pool. Non-conflicting systems can run in parallel.
+    MultiThreaded,
+}
+
+#[derive(Default)]
+pub(super) struct SystemSchedule {
+    pub(super) systems: Vec<RefCell<BoxedSystem>>,
+    pub(super) system_conditions: Vec<RefCell<Vec<BoxedCondition>>>,
+    pub(super) set_conditions: Vec<RefCell<Vec<BoxedCondition>>>,
+    pub(super) system_ids: Vec<NodeId>,
+    pub(super) set_ids: Vec<NodeId>,
+    pub(super) system_deps: Vec<(usize, Vec<usize>)>,
+    pub(super) sets_of_systems: Vec<FixedBitSet>,
+    pub(super) sets_of_sets: Vec<FixedBitSet>,
+    pub(super) systems_of_sets: Vec<FixedBitSet>,
+}
+
+impl SystemSchedule {
+    pub const fn new() -> Self {
+        Self {
+            systems: Vec::new(),
+            system_conditions: Vec::new(),
+            set_conditions: Vec::new(),
+            system_ids: Vec::new(),
+            set_ids: Vec::new(),
+            system_deps: Vec::new(),
+            sets_of_systems: Vec::new(),
+            sets_of_sets: Vec::new(),
+            systems_of_sets: Vec::new(),
+        }
+    }
+}
+
+// SAFETY: MultiThreadedExecutor does not alias RefCell instances
+unsafe impl Sync for SystemSchedule {}
```

Comment:
For further safety, please leave a comment that `SystemSchedule`, the type itself, to ensure that:
 
 - `MultiThreadedExecutor` is the only type that uses `SystemSchedule` across multiple threads.
 - `MultiThreadedExecutor` cannot alias any of the RefCells.
 - `SystemSchedule` is not made pub in any way.

Reply:
Is it possible to use a SyncCell to make the type Sync?

Reply:
I'm thinking that it would probably be best to use SyncUnsafeCell and then use the unsafe functions on it where needed. That way the user of the type knows that they need to uphold the invariant when they use it rather than it just falling into UB if they don't see the comment on the type.

Reply:
This comment needs to be added above before we can merge.

Reply:
> I'm thinking that it would probably be best to use SyncUnsafeCell.

Agreed. I'll swap replace `RefCell` use with `SyncUnsafeCell`.